src/components/time-picker/time-picker.tsx

```diff
@@ -0,0 +1,382 @@
+        const nextSegment = segments[index + step];
+        if (nextSegment) {
+            return this.select(nextSegment);
+        }
+        return false;
+    }
+
+    private updateSegmentValue(name: Segment, value: number | Ampm) {
+        const {hh, mm, ampm} = this.state;
+        this.setState({
+            [name as any]: value,
+            inputValue: segmentsToInputValue({
+                hh, mm, ampm,
+                [name]: value
+            })
+        }, () => {
+            this.input!.focus();
+            this.select(name);
+            this.commit();
+        });
+    }
+
+    private toggleAmpm() {
+        this.updateSegmentValue('ampm', 1 - this.state.ampm);
+    }
+
+    private changeValue(step: number) {
+        const name = this.state.currentSegment;
+        if (!isTimeSegment(name)) {
+            return this.toggleAmpm();
+        }
+        const {ampm} = this.state;
+        const newValue = (this.state[name] || 0) + step;
+
+        if (isValidValue(newValue, name, ampm)) {
+            this.updateSegmentValue(name, newValue);
+        }
+    }
+
+    private createStepperHandler = (step: number) => () => this.changeValue(step);
+
+    private onInputMouseDown = (e: any) => {
+        // when user select text with double click
+        if (e.detail > 1) {
+            e.preventDefault();
+        }
+    }
+
+    private onInputClick = (e: React.SyntheticEvent<HTMLInputElement>) => {
+        const cursorPosition = e.currentTarget.selectionStart;
+        let key: Segment;
+        let currentSegment: Segment = 'hh';
+
+        for (key in selectionIndexes) {
+            if (cursorPosition >= selectionIndexes[key][0] && cursorPosition <= selectionIndexes[key][1]) {
+                currentSegment = key;
+            }
+        }
+        if (currentSegment === 'ampm') {
+            this.toggleAmpm();
+        } else {
+            this.select(currentSegment);
+        }
+    }
+
+    private onInputChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
+        const {format} = this.props;
+        const {ampm} = this.state;
+        let {value} = e.currentTarget;
+        if (value.length === 1) {
+            value = defaultValue(format!, value);
+        }
+        const match = value.match(/^(\d{1,2}):(\d{1,2})(?:\s(AM|PM))?$/);
+        if (!match) {
+            return;
+        }
+        const hh = Number(match[1]);
+        const mm = Number(match[2]);
+
+        if (
+            !isValidValue(hh, 'hh', ampm) ||
+            !isValidValue(mm, 'mm', ampm) ||
+            (format === 'ampm' && !match[3])
+        ) {
+            return;
+        }
+
+        const input = this.input!;
+        const cursorPosition = input.selectionStart;
+        const currentSegment = this.state.hh !== hh ? 'hh' : 'mm';
+        const updatedValue = this.state.hh !== hh ? hh : mm;
+        const shouldWaitForInput = isValidValue(updatedValue * 10, currentSegment, ampm);
```

Comment:
`isValidValue` is not good enough? I may add description but there is no need for separate function. This function also used for `stepper` to make disable states for buttons with the same principle (`+1` and `-1` used instead of `* 10`)